### PR TITLE
Restore most serialization implementations in coreclr

### PR DESCRIPTION
**Pull request by stephentoub**
I audited all uses of FEATURE_SERIALIZATION in coreclr and removed most of the ifdef'ing (this required a few minor changes beyond removing the ifdefs in a few places).  The remaining uses of FEATURE_SERIALIZATION are primary related to actually implementing formatters, serializing CAS-related types, and serializing a few types I've left ifdef'd (e.g. ExecutionContext) but that we can revisit later if needed.

I will add tests for all of this to corefx, but that's blocked on exposing the coreclr synchronization primitives via type forwards rather than using the versions defined in corefx, and that's blocked on an issue building partial facades against mscorlib that @weshaggard is working on.

NOTE: We should not merge this until that work is done, both so that we can test this before merging it and so that we're not enabling serialization of types that will have faulty implementations due to the formatter not recognizing the interfaces/attributes they're using to implement custom serialization.

cc: @jkotas, @danmosemsft, @weshaggard, @joperezr

### CODE REVIEW

**Comment by stephentoub** [comment]
I missed the Unix-specific implementations of the globalization stuff (that's the cause of the CI failures); I'll add those.

File: src/mscorlib/src/System/Diagnostics/Contracts/ContractsBCL.cs [hunk 1]
```diff
@@ -202,11 +202,6 @@ public void SetUnwind()
         }
     }
 
-#if FEATURE_SERIALIZATION
-    [Serializable]
-#else
-    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1032:ImplementStandardExceptionConstructors")]
-#endif
     [SuppressMessage("Microsoft.Design", "CA1064:ExceptionsShouldBePublic")]
```
Comment (jkotas) [inline src/mscorlib/src/System/Diagnostics/Contracts/ContractsBCL.cs]: Is the removal of `[Serializable]` on `ContractException` intentional? I think it should stay.

Comment (stephentoub) [inline src/mscorlib/src/System/Diagnostics/Contracts/ContractsBCL.cs]: Unintentional fat-fingering.  Fixed.

**Comment by stephentoub** [comment]
Test Windows_NT x86 ryujit Checked Build and Test please (failure to restore package)
Test Linux ARM Emulator Cross Release Build please (build timeout)

**Comment by danmosemsft** [comment]
@sepidehMS

**Comment by stephentoub** [comment]
Test Ubuntu x64 Checked Build and Test please

**Comment by weshaggard** [comment]
@stephentoub are still waiting on something from me for this PR?

**Comment by stephentoub** [comment]
> are still waiting on something from me for this PR?

Nope, I plan to merge this once CI completes.

File: src/mscorlib/corefx/System/Globalization/Calendar.cs [hunk 2]
```diff
@@ -29,7 +30,7 @@ namespace System.Globalization
     // The calculation of hour/minute/second is moved to Calendar from GregorianCalendar,
     // since most of the calendars (or all?) have the same way of calcuating hour/minute/second.
 
-
+    [Serializable]
```
Comment (weshaggard) [inline src/mscorlib/corefx/System/Globalization/Calendar.cs]: How did you find all the things that need to be annotated? As I assume the source diff would be difficult. 

We will also need to get this change into corert repo.

Comment (stephentoub) [inline src/mscorlib/corefx/System/Globalization/Calendar.cs]: I searched for all serializable types in globalization and then went through and annotated the corresponding ones here.

**Comment by danmosemsft** [comment]
Hey @stephentoub should this activeissue be removed now? Or changed to an actual new issue?
```
C:\git\corefx\src\System.Runtime.Serialization.Formatters\tests\SerializationInfoTests.cs:
        [ActiveIssue("https://github.com/dotnet/coreclr/pull/6423")]
        [Fact]
        public void NegativeValueNotFound()
```

**Comment by stephentoub** [comment]
> should this activeissue be removed now?

If it passes, then yes ;)

**Comment by danmosemsft** [comment]
@Jiayili1 @shimingsg could one of you please do the same thing here as requested in https://github.com/dotnet/coreclr/pull/7966#issuecomment-284019823
thanks!

**Pull request merged**